flowauth/backend/flowauth/roles.py

```diff
@@ -0,0 +1,185 @@
+
+
+@blueprint.route("/", methods=["POST"])
+@login_required
+@admin_permission.require(http_exception=401)
+def add_role():
+    json = request.get_json()
+    json["latest_token_expiry"] = datetime.datetime.strptime(
+        json["latest_token_expiry"], "%Y-%m-%dT%H:%M:%S.%fZ"
+    )
+    server = Server.query.filter(Server.id == json["server_id"]).first_or_404()
+    role_scopes = Scope.query.filter(Scope.id.in_(json["scopes"])).all()
+    for scope in role_scopes:
+        _validate_scope_server(scope, server)
+    try:
+        role_users = User.query.filter(User.id.in_(json["users"])).all()
+    except KeyError:
+        role_users = []
+
+    if json["latest_token_expiry"] > server.latest_token_expiry:
+        raise InvalidUsage("Role cannot exist past latest token in server")
+    if int(json["longest_token_life_minutes"]) > server.longest_token_life_minutes:
+        raise InvalidUsage("Role cannot have a maximum lifetime greater than server")
+
+    new_role = Role(
+        name=json["name"],
+        scopes=role_scopes,
+        users=role_users,
+        server=server,
+        latest_token_expiry=json["latest_token_expiry"],
+        longest_token_life_minutes=json["longest_token_life_minutes"],
+    )
+    db.session.add(new_role)
+    try:
+        db.session.commit()
+    except IntegrityError as e:
+        db.session.rollback()
+        raise e
+    current_app.logger.info(f"Created role {new_role.name}")
+    return get_role(new_role.id)
+
+
+# NOTES FOR REVIEW: I wondered whether to have this merge lists of users and scopes
+# when updated, but in the end I think that it's cleaner to have that on the frontend
+# and make an explicit decision
+@blueprint.route("/<role_id>", methods=["PATCH"])
+@login_required
+@admin_permission.require(http_exception=401)
+def edit_role(role_id):
+    edits = request.get_json()
+    role = Role.query.filter(Role.id == role_id).first_or_404()
+    server = role.server
+    for key, value in edits.items():
+        if key == "id":
+            current_app.logger.warning("Cannot change role ID; ignoring")
+            continue
+        elif key == "users":
+            value = [User.query.filter(User.id == uid).first() for uid in value]
+        elif key == "scopes":
+            value = [Scope.query.filter(Scope.id == sid).first() for sid in value]
+            for scope in value:
+                _validate_scope_server(scope, server)
+        elif key == "latest_token_expiry":
+            value = datetime.datetime.strptime(value, "%Y-%m-%dT%H:%M:%S.%fZ")
+            if value > server.latest_token_expiry:
+                raise InvalidUsage("Role cannot exist past latest token in server")
+        elif key == "longest_token_lifetime_minutes":
+            if value > server.longest_token_life_minutes:
+                raise InvalidUsage(
+                    "Role cannot have a maximum lifetime greater than server"
+                )
+        elif key not in ["name", "server_id"]:
```

Comment:
Does the ability to change the server id mean that you can create a role with a later expiry than should be allowed?

Reply:
And presumably invalid scopes, now I think about it.

Reply:
As in, on server A (exp. Now+10) you create role 1 (exp. Now+10) and then move it to server B (exp. Now+1)?
Yes. I think that being able to move roles between servers is also more trouble than it's worth, so 'server_id' will come out.